test/functional/test_framework/mininode.py

```diff
@@ -524,6 +551,33 @@ def close(self, timeout=10):
         # Safe to remove event loop.
         NetworkThread.network_event_loop = None
 
+    @classmethod
+    def listen(cls, p2p, callback, port=None, addr=None, idx=0):
+        if port is None:
+            assert idx >= 0 and idx < MAX_NODES
+            port = p2p_port(MAX_NODES-idx)
+        if addr is None:
+            addr = '127.0.0.1'
+        coroutine = cls.create_server(addr, port, callback, p2p)
+        cls.network_event_loop.call_soon_threadsafe(cls.network_event_loop.create_task, coroutine)
+
+    @classmethod
+    def peer_listener(cls, addr, port):
+        def pl():
+            x = cls.protos.get((addr,port))
+            cls.protos[(addr,port)] = None
+            return x
+        return pl
+
+    @classmethod
+    async def create_server(cls, addr, port, callback, proto):
+        if (addr, port) not in cls.listeners:
+            listener = await cls.network_event_loop.create_server(cls.peer_listener(addr, port), addr, port)
+            logger.debug("Listening server on %s:%d should be started" % (addr, port))
```

Comment:
I looked into the flaky test a bit.

Locally, I see this:

	2020-08-27T06:35:37.249000Z TestFramework.mininode (DEBUG): Listening for Bitcoin Node: 0:
	2020-08-27T06:35:37.250000Z TestFramework.mininode (DEBUG): Listening server on 127.0.0.1:15398 should be started

However, in the CI logs, the "Listening server on ..." never appears:

	2020-08-25T06:50:18.202000Z TestFramework.mininode (DEBUG): Listening for Bitcoin Node: 0: 
	2020-08-25T06:51:18.213000Z TestFramework.utils (ERROR): wait_until() failed.
	
It looks like on the CI servers, the `cls.network_event_loop.create_server` call didn't work:

	Traceback (most recent call last):
	  File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/mininode.py", line 575, in create_server
	    listener = await cls.network_event_loop.create_server(cls.peer_listener(addr, port), addr, port)
	  File "/usr/lib/python3.8/asyncio/base_events.py", line 1463, in create_server
	    raise OSError(err.errno, 'error while attempting '
	OSError: [Errno 98] error while attempting to bind on address ('127.0.0.1', 15020): address already in use

Unfortunately, I have **no idea** why this port is already in use on the CI server—the CI logs make no other mention of port 15020 before this error.

- Can we guarantee that port 15020 isn't used by any other processes on the CI server?
- If not, would it make sense to retry the `create_server` with different ports on failure?

Reply:
thank you very much for looking into the flaky test :) you are a brave soul! hahaha

with ~some~ a lot of help from @ajtowns, my best understanding of what was probably happening here is an off-by-one error. the port allocation is intended to have 12 unique slots for each test, allowing test runner to run multiple tests in parallel. previously, the idx was starting at 0, so p2p_port was being passed (MAX_NODES), which could lead to a port conflict with another test. to fix it, I'm starting the idx at 1 & have added the validations to check the proper range. let's see how CI does 🤞🏽

to be more specific, the port conflict I'm referring to would occur if a test had pid `x` and invoked `p2p_port(12)`, and the next test had pid `x+1` and invoked `p2p_port(0)`.